### PR TITLE
Introduce create storage pools action

### PR DESCRIPTION
**Pull request by KKoukiou**

### CODE REVIEW

**Comment by andreasn** [comment]
Can you make this inline instead?
So that the Name field become highlighted and the parts about XML feels like an implementation detail, since I'm not editing XML here.

screenshot from 2018-11-21 15-44-58

File: pkg/machines/components/storagePools/createStoragePoolDialog.jsx [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+                </Grid.Col>
+            </FormGroup>
+        );
+    return null;
+};
+
+const StoragePoolTypeRow = ({ onValueChanged, dialogValues }) => {
+    const poolTypes = [
+        { type: 'dir', detail: 'Filesystem Directory' },
+        { type: 'netfs', detail: 'Network Exported Directory' },
```
Comment (andreasn) [inline pkg/machines/components/storagePools/createStoragePoolDialog.jsx]: Is this the same as NFS or something different?

File: pkg/machines/components/storagePools/createStoragePoolDialog.jsx [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+                </Grid.Col>
+            </FormGroup>
+        );
+};
+
+const StoragePoolAutostartRow = ({ onValueChanged, dialogValues }) => {
+    return (
+        <FormGroup controlId='autostart'>
+            <Grid.Col componentClass={Form.ControlLabel} sm={3}>
+                {_("Autostart")}
```
Comment (andreasn) [inline pkg/machines/components/storagePools/createStoragePoolDialog.jsx]: What does Autostart mean in this case? Does it start when the libvirtd service start, or something else?

Comment (KKoukiou) [reply pkg/machines/components/storagePools/createStoragePoolDialog.jsx]: It determines whether the pool is automatically started at boot time.

Comment (andreasn) [reply pkg/machines/components/storagePools/createStoragePoolDialog.jsx]: How about something like this for a bit more clarity:
"**Startup** [x] Start pool when host boots"

File: pkg/machines/components/storagePools/createStoragePoolDialog.jsx (hunk 1 above)
Comment (KKoukiou) [reply pkg/machines/components/storagePools/createStoragePoolDialog.jsx]: Libvirt under nfs category has some other types like GFS. Anyway, I think we are interested only in nfs in cockpit, so I can rename the netfs to nfs.

**Comment by andreasn** [comment]
Maybe this is due to patternfly-react, but I get a big gap to the right to the input controls. I haven't seen this elsewhere, so I wonder what's causing it.
screenshot from 2018-11-21 15-44-58

**Comment by andreasn** [comment]
Checked in the inspector now, and I think it's because it's using col-sm-6 instead of col-sm-9

**Comment by KKoukiou** [comment]
@andreasn The gap it's because I use col-sm-3 class for the labels and col-sm-6 class for the input fields. So the reminder 3 columns in the modal grid right are unused.
You want to suggest some other setup?

**Comment by KKoukiou** [comment]
@andreasn right, don't you think col-sm-9 is kinda too much for the input fields?

**Comment by andreasn** [comment]
Hm, yeah, maybe it's a bit intense.
I think the only part that would benefit from col-sm-9 would be very deep paths.
screenshot from 2018-11-21 16-14-28

**Comment by andreasn** [comment]
But! We can also make the dialog a bit slimmer.
We do this in the system time dialog and some other places.

**Comment by andreasn** [comment]
Here is how it would look if the dialog was 500px wide:
screenshot from 2018-11-21 17-13-16

**Comment by KKoukiou** [comment]
@and thanks, I 'll make sure to make it smaller. Either overwrite css or check if Modal from patternfly is configurable in matters of size.

File: pkg/machines/libvirt-dbus.es6 [hunk 3]
```diff
@@ -517,7 +517,8 @@ LIBVIRT_DBUS_PROVIDER = {
                         props.available = poolInfo[0][3];
 
                         dispatch(updateOrAddStoragePool(Object.assign({}, dumpxmlParams, props)));
-                        dispatch(getStorageVolumes({ connectionName, poolName: dumpxmlParams.name }));
+                        if (props.active)
+                            dispatch(getStorageVolumes({ connectionName, poolName: dumpxmlParams.name }));
```
Comment (Antique) [inline pkg/machines/libvirt-dbus.es6]: This change seems to be unrelated to this pull request.

File: pkg/machines/libvirt-dbus.es6 [hunk 4]
```diff
@@ -268,6 +270,31 @@ LIBVIRT_DBUS_PROVIDER = {
...
+        host,
+        autostart,
+    }) {
+        const poolXmlDesc = getPoolXML({name, type, source, target, host});
+        let storagePoolPath;
+
+        return (dispatch) => call(connectionName, '/org/libvirt/QEMU', 'org.libvirt.Connect', 'StoragePoolDefineXML', [poolXmlDesc, 0], TIMEOUT)
+                .then(poolPath => {
+                    storagePoolPath = poolPath;
+                    return call(connectionName, storagePoolPath[0], 'org.libvirt.StoragePool', 'Create', [Enum.VIR_STORAGE_POOL_CREATE_NORMAL], TIMEOUT);
```
[the remark this answers is not in the document]
Comment (andreasn) [reply pkg/machines/libvirt-dbus.es6]: Sounds a little edge-casey to me.
Why wouldn't you just start the NFS server first?

[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/libvirt-dbus.es6]: I understand the usecase but it's indeed kinda edgy and I would rather not flood the modal with 'not-clear-what-they-do' options to support such scenarios.
At least until the need appears.

File: pkg/machines/libvirt-dbus.es6 (hunk 3 above)
Comment (KKoukiou) [reply pkg/machines/libvirt-dbus.es6]: It's separate commit, check commits.
I tend to put multiple commits into a PR, that sometimes might seem unrelated, but posting a single line PR just for a small fix like this is a bit of overkill since all tests will be triggered.

File: pkg/machines/components/storagePools/createStoragePoolDialog.jsx (hunk 2 above)
Comment (KKoukiou) [reply pkg/machines/components/storagePools/createStoragePoolDialog.jsx]: I like the help message, but if you don't mind I 'd rather keep the label name to autostart since it's used for other resources as well, and it matches the libvirt terminology with which some people might already be familiar with.

File: pkg/machines/libvirt-dbus.es6 (hunk 4 above)
[the remark this answers is not in the document]
Comment (Antique) [reply pkg/machines/libvirt-dbus.es6]: You might not be connected to the correct network at the moment, it's not always the case that the NFS server is not started, it might be running, that's why I used "not accessible".  This would mimic the same logic as we have for domains, you can also define domain without starting it.  I agree that it will not be used that often, just thought that it would be nice feature to have.

[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/libvirt-dbus.es6]: Let's add the basic functionality first and we can consider such advanced options for future RFE.

File: pkg/machines/libvirt-dbus.es6 (hunk 3 above)
Comment (Antique) [reply pkg/machines/libvirt-dbus.es6]: I know that it's separate commit, that's exactly why I'm asking, whether it's a mistake or intentional.

Well, since you've used that as reasoning I cannot resist to point out that in that case the whole process is broken :) you should be able to merge simple fixes without triggering all tests.

File: pkg/machines/libvirt-dbus.es6 (hunk 4 above)
[the remark this answers is not in the document]
Comment (Antique) [reply pkg/machines/libvirt-dbus.es6]: Sure, I'm OK with that.

File: pkg/machines/libvirt-dbus.es6 (hunk 3 above)
Comment (KKoukiou) [reply pkg/machines/libvirt-dbus.es6]: Theoretically we can trigger parts of the test-suite as well but we don't do that and it works perfect so far. And having some fixes for the same package inside a PR that was meant to mainly do something else is not a great deal - it makes things quicker without making the commit history dirty.

File: pkg/machines/xmlCreator.es6 [hunk 5]
```diff
@@ -45,3 +45,36 @@ export function getVolumeXML(volumeName, size, format, target) {
...
+        let dirElem = doc.createElement('dir');
+        dirElem.setAttribute('path', source.dir);
+        sourceElem.appendChild(dirElem);
+        if (host) {
+            let hostElem = doc.createElement('host');
+            hostElem.setAttribute('name', host);
+            sourceElem.appendChild(hostElem);
+        }
+    }
+    poolElem.appendChild(sourceElem);
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/xmlCreator.es6]: Done

File: pkg/machines/xmlCreator.es6 [hunk 6]
```diff
@@ -45,3 +45,36 @@ export function getVolumeXML(volumeName, size, format, target) {
...
+    poolElem.appendChild(nameElem);
+
+    let targetElem = doc.createElement('target');
+    let pathElem = doc.createElement('path');
+    pathElem.appendChild(doc.createTextNode(target));
+    targetElem.appendChild(pathElem);
+    poolElem.appendChild(targetElem);
+
+    let sourceElem = doc.createElement('source');
+    if (source.dir) {
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/xmlCreator.es6]: Done

**Comment by andreasn** [comment]
Awesome work. Thanks for fixing this so quickly!
I haven't seen the little X symbols elsewhere in Cockpit, but maybe that comes automatically when you use patternfly-react?

File: pkg/machines/components/storagePools/createStoragePoolDialog.jsx (hunk 2 above)
Comment (andreasn) [reply pkg/machines/components/storagePools/createStoragePoolDialog.jsx]: What about "**Startup** [x] Autostart pool on boot"
That way it explains a little better, but still keeps the familiar term **Autostart**.
Since we're primarily targeting new linux admins (including senior Windows admins), we need something that works for people who are not previously familiar with libvirt.

It also places the checkbox on the left of the label it's connected to, and that's the way checkboxes works everywhere else. Not super-happy about the label **Startup**, but it was the best I could come up with.

**Comment by KKoukiou** [comment]
> Awesome work. Thanks for fixing this so quickly!
> I haven't seen the little X symbols elsewhere in Cockpit, but maybe that comes automatically when you use patternfly-react?

I added the X explicitly, it's indeed a patternfly-react component but it can be removed. Should I remove it?

**Comment by andreasn** [comment]
> I added the X explicitly, it's indeed a patternfly-react component but it can be removed. Should I remove it?

Maybe best. It would be most consistent with how we do the error state elsewhere in Cockpit.

File: test/common/testlib.py [hunk 7]
```diff
@@ -245,6 +245,17 @@ def key_press(self, keys, modifiers=0):
 
             self.cdp.invoke("Input.dispatchKeyEvent", **args)
 
+    def select_from_dropdown(self, selector, value):
+        button_text_selector = "{0} button span:nth-of-type(1)".format(selector)
+
+        self.wait_visible(selector)
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply test/common/testlib.py]: Done

File: pkg/machines/components/storagePools/createStoragePoolDialog.css [hunk 8]
```diff
@@ -0,0 +1,3 @@
+.modal-dialog {
+    width: 500px;
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/components/storagePools/createStoragePoolDialog.css]: Oh, good point, @andreasn are you ok with this change?

File: pkg/machines/components/storagePools/createStoragePoolDialog.jsx [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+import cockpit from 'cockpit';
+
+import './createStoragePoolDialog.css';
+
+const _ = cockpit.gettext;
+
+const StoragePoolConnectionRow = ({ onValueChanged, dialogValues, loggedUser }) => {
+    let connectionUris = [
+        <Select.SelectEntry data={LIBVIRT_SYSTEM_CONNECTION}
+                            key={LIBVIRT_SYSTEM_CONNECTION}>{_("QEMU/KVM System connection")}
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/components/storagePools/createStoragePoolDialog.jsx]: I 'll do that in followup, thanks :)

File: pkg/machines/components/storagePools/createStoragePoolDialog.jsx [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+                </HelpBlock> }
+            </Grid.Col>
+        </FormGroup>
+    );
+};
+
+const StoragePoolTypeRow = ({ onValueChanged, dialogValues }) => {
+    const poolTypes = [
+        { type: 'dir', detail: 'Filesystem Directory' },
+        { type: 'netfs', detail: 'Network Exported Directory' },
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/components/storagePools/createStoragePoolDialog.jsx]: Done.

File: pkg/machines/components/storagePools/createStoragePoolDialog.jsx [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+            </Grid.Col>
+            <Grid.Col sm={9}>
+                <Select.Select id='storage-pool-dialog-type'
+                               initial={dialogValues.type}
+                               onChange={value => onValueChanged('type', value)}>
+                    { poolTypes
+                            .map(pool => {
+                                return (
+                                    <Select.SelectEntry data={pool.type} key={pool.type}>
+                                        {`${pool.type == 'netfs' ? 'nfs' : pool.type} : ${pool.detail}`}
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/components/storagePools/createStoragePoolDialog.jsx]: I am ok with not showing the abbreviation but only the detail, @andreasn are you ok with this?

File: pkg/machines/components/storagePools/createStoragePoolDialog.css (hunk 8 above)
[the remark this answers is not in the document]
Comment (andreasn) [reply pkg/machines/components/storagePools/createStoragePoolDialog.css]: Everywhere else I've seen, the dialogs are defined in pixels, but I'm not 100% sure.
@garrett do you know what's best practice here?

File: pkg/machines/components/storagePools/createStoragePoolDialog.jsx [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+                </Select.Select>
+            </Grid.Col>
+        </FormGroup>
+    );
+};
+
+const StoragePoolTargetRow = ({ onValueChanged, dialogValues }) => {
+    const validationState = dialogValues.target.length == 0 && dialogValues.validationFailed.target ? 'error' : undefined;
+
+    if (['dir', 'netfs'].includes(dialogValues.type))
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/components/storagePools/createStoragePoolDialog.jsx]: Done. And we can then enable the https://eslint.org/docs/2.0.0/rules/curly "multi-line" rule.

File: pkg/machines/components/storagePools/createStoragePoolDialog.jsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (andreasn) [reply pkg/machines/components/storagePools/createStoragePoolDialog.jsx]: This is tricky.
People who want to share things over NFS will look for that string specifically, or possibly the string **Network File System**. It is also the language that Windows Server uses.
https://docs.microsoft.com/en-us/windows-server/storage/nfs/deploy-nfs

Is **Network Export Directory** language that is unique to KVM/QEMU/libvirtd?

NFS is also what oVirt uses, as far as I can see.
screenshot from 2018-11-26 12-18-44

Even though it's a bit redundant, I would keep both :/

[the remark this answers is not in the document]
Comment (andreasn) [reply pkg/machines/components/storagePools/createStoragePoolDialog.jsx]: Ah, right. Yeah, let's go for **Network File System** as the detail string, and show only the detail string then.

File: test/verify/check-machines-dbus [hunk 13]
```diff
@@ -224,6 +224,155 @@ class TestMachinesDBus(machineslib.TestMachines):
...
+                self.autostart = autostart
+                self.xfail = xfail
+                self.used_names = []
+
+            def execute(self):
+                self.open()
+                self.fill()
+                self.create()
+                if not self.xfail:
+                    self.verify_dialog()
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply test/verify/check-machines-dbus]: Done.
I have the boolean xfail now, and the xfail_error string which matches the errors from the backend.
We do also some checks for validity in react code, so for these, we pass xfail=True but xfail_error is not set.

File: pkg/machines/components/storagePools/createStoragePoolDialog.css (hunk 8 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply pkg/machines/components/storagePools/createStoragePoolDialog.css]: FTR, this wasn't a veto. I know relatively little about our best practices here, and if we do the same in other places, it's fine.

File: pkg/machines/libvirt-dbus.es6 [hunk 14]
```diff
@@ -969,6 +977,28 @@ function startEventMonitor(dispatch, connectionName, libvirtServiceName) {
...
+        (path, iface, signal, args) => {
+            let objPath = args[0];
+            let eventType = args[1];
+
+            switch (eventType) {
+            case Enum.VIR_STORAGE_POOL_EVENT_DEFINED:
+            case Enum.VIR_STORAGE_POOL_EVENT_STARTED:
+            case Enum.VIR_STORAGE_POOL_EVENT_STOPPED:
+            case Enum.VIR_STORAGE_POOL_EVENT_CREATED:
+                dispatch(getStoragePool({connectionName, id:objPath}));
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/libvirt-dbus.es6]: VIR_STORAGE_POOL_EVENT_UNDEFINED && VIR_STORAGE_POOL_EVENT_DELETED are now ignored. I haven't implemented the reducers to remove existing storage pools yet, I have to implement it. So, if we receive on of these two events for now, nothing will happen, I have to implement the removeStoragePool reducer. Which I 'll do asap, it kinda a bug gap indeed.

File: pkg/machines/components/storagePools/createStoragePoolDialog.css (hunk 8 above)
[the remark this answers is not in the document]
Comment (garrett) [reply pkg/machines/components/storagePools/createStoragePoolDialog.css]: In this case, it would be better to size with `rem`, which is the root em (that is: the default font size, instead of whatever font size might be on the element). However, PatternFly uses `px` sizes for nearly everything, so it's not entirely wrong, especially as a web `px` is not necessarily a device pixel (in other words; a `px` _may_ scale between devices, browsers, or browser overrides). Using `em`/`rem` does size based on font size, however, so you do have predictable amounts of text.

For modals in particular, PatternFly does inherit from Bootstrap, which does have `modal-lg` and `modal-sm` add-on classes for `modal-dialog`, but `modal-sm` is set for a width of `300px`, whereas standard modals are set to `600px`.

_**TL;DR**: Using `rem` might be better but, sure, this works._ :+1:

[the remark this answers is not in the document]
Comment (garrett) [reply pkg/machines/components/storagePools/createStoragePoolDialog.css]: HOWEVER: You _might_ want to not change `.modal-dialog` to 500px, but key it off of something like `.ct-modal-storage` and set that to have width 500px and include that class in addition to `modal-dialog` however. (That is, `className="modal-dialog ct-modal-storage"`)
